File: raw_scripts/V5/clus.py
#!/usr/bin/env python
import sys
import hdbscan
import numpy as np
# import matplotlib
# matplotlib.use('Agg')
# import matplotlib.pyplot as plt
# import seaborn as sns
# import argparse
# import logging
import os
def hdbscan_clustering(file_name,mat_file,minpts,minsamps,baseline_samp_label,treated_samp_label,cores):
    path = os.getcwd()
    os.chdir(path)
    d = np.genfromtxt(mat_file, delimiter=',')
    clusterer = hdbscan.HDBSCAN(min_cluster_size = minpts, min_samples = minsamps, core_dist_n_jobs = cores).fit(d)
    # color_palette = sns.color_palette('deep', 8)
    # cluster_colors = [color_palette[x] if x >= 0
    #                 else (0.5, 0.5, 0.5)
    #                 for x in clusterer.labels_]
    # cluster_member_colors = [sns.desaturate(x, p) for x, p in
    #                         zip(cluster_colors, clusterer.probabilities_)]
    np.savetxt('clus.%s.%s.%s.txt' % (file_name, minpts, minsamps), clusterer.labels_.astype(int), fmt = '%i')
    # plt.scatter(*d.T, s=50, linewidth=0, c=cluster_member_colors, alpha=0.25)
    # plt.xlabel(baseline_samp_label)
    # plt.ylabel(treated_samp_label)
    # plt.savefig('clus.%s.%s.%s.png' % (file_name, minpts, minsamps), transparent=True)

File: raw_scripts/V5/clus.R
#!/usr/bin/env Rscript
clus <- function(output_file_name,
                 out_dir,
                 matrix_file,
                 baseline_samp_label,
                 treated_samp_label,
                 minpts,
                 minsamps,
                 cores) {
  install_HDBSCAN <-
    function(...,
             envname = "r-HBDSCAN",
             new_env = identical(envname, "r-HBDSCAN")) {

      if(new_env && virtualenv_exists(envname))
        virtualenv_remove(envname)

      reticulate::py_install(packages = c("hdbscan","numpy","matplotlib","seaborn"), envname = envname, ...)

      .onLoad <- function(...) {
        use_virtualenv("r-HDBSCAN", required = FALSE)
      }
    }
  reticulate::source_python(system.file("python", "clus.py", package = "ChIPbinner"))
  # file parameters
  output_file_name <- paste0(output_file_name)
  # output directory
  setwd(paste0(out_dir))
  # samples info
  baseline_samp_label <- paste0(baseline_samp_label)
  treated_samp_label <- paste0(treated_samp_label)
  # input file
  mat_file <- paste0(matrix_file)
  # algorithm parameters
  minpts <- as.integer(paste0(minpts))
  minsamps <- as.integer(paste0(minsamps))
  # number of parallel cores to use
  cores <- as.integer(paste0(cores))
  # out_dir
  hdbscan_clustering(
    file_name = output_file_name,
    mat_file = mat_file,
    minpts = minpts,
    minsamps = minsamps,
    baseline_samp_label = baseline_samp_label,
    treated_samp_label = treated_samp_label,
    cores = cores
  )
  print("Density-based clusters generated using HDBSCAN!")
}
